src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNet/AspNetMvcIntegration.cs

```diff
@@ -90,22 +93,52 @@ public static Scope CreateScope(ControllerContextStruct controllerContext)
                     }
                 }
 
+                string routeUrl = route?.Url;
+                string areaName = (routeValues?.GetValueOrDefault("area") as string)?.ToLowerInvariant();
+                string controllerName = (routeValues?.GetValueOrDefault("controller") as string)?.ToLowerInvariant();
+                string actionName = (routeValues?.GetValueOrDefault("action") as string)?.ToLowerInvariant();
+
+                if (newResourceNamesEnabled && string.IsNullOrEmpty(resourceName) && !string.IsNullOrEmpty(routeUrl))
+                {
+                    resourceName = $"{httpMethod} /{routeUrl.ToLowerInvariant()}";
+                }
+
                 if (string.IsNullOrEmpty(resourceName) && httpContext.Request.Url != null)
                 {
                     var cleanUri = UriHelpers.GetRelativeUrl(httpContext.Request.Url, tryRemoveIds: true);
                     resourceName = $"{httpMethod} {cleanUri.ToLowerInvariant()}";
                 }
 
-                string areaName = (routeValues?.GetValueOrDefault("area") as string)?.ToLowerInvariant();
-                string controllerName = (routeValues?.GetValueOrDefault("controller") as string)?.ToLowerInvariant();
-                string actionName = (routeValues?.GetValueOrDefault("action") as string)?.ToLowerInvariant();
-
                 if (string.IsNullOrEmpty(resourceName))
                 {
                     // Keep the legacy resource name, just to have something
                     resourceName = $"{httpMethod} {controllerName}.{actionName}";
                 }
 
+                // Fail safe to catch templates in routing values
+                resourceName =
+                    resourceName
+                       .Replace("{area}", areaName)
+                       .Replace("{controller}", controllerName)
+                       .Replace("{action}", actionName);
+
+                if (newResourceNamesEnabled && !wasAttributeRouted && routeValues is not null && route is not null)
+                {
+                    // Remove unused parameters from conventional route templates
+                    // Don't bother with routes defined using attribute routing
```

Comment:
IMO, it's a difference in philosophy between the two routing types (and will have a slight performance benefit to not doing the extra work). 

Conventional routes (in particular the _default_ route) often have a placeholder `{id?}` (or similar) parameter, as they're trying to match a single route to multiple actions. So displaying `/home/index/{id}` wouldn't make any sense when the URL is `/home/index` and the  action is `Index()`.

In contrast, attribute routes are 1-1 tied to an action. So stripping off the (in my experience, rare) optional parameter would give the appearance of two different endpoints when they aren't,  so doesn't make sense to me I think. For example, given the action:

```csharp
[HttpGet, Route("api/optional-delay/{value}")]
public IActionResult OptionalDelay(int value = 5) { /* */}
```

I think it's more beneficial to have `/api/optional-delay` and `/api/optional-delay/5` both use the route `api/optional-delay/{value}`, instead of two separate routes for the same endpoint. As I say, the nature of the 1-1 mapping means optional parameters are pretty rare in these cases anyway.

Now, you could make an argument that for the conventional routing, if the user calls `/home/index/123`, we will use the route `/home/index/{id}` and they will still be routed to the same `Index()` action. In that case we have two resources with different resource name routes (one with `{id}` and one without), but in my experience this is _very_ rarely used in conventional routing, and more often will point to a bug. And the alternative is never stripping the optional routes, will just look weird most of the time. 

`</essay>`

Reply:
Your logic makes sense to me. Thanks for the essay 😉